File: app/src/main/java/com/spproject/audiowatermark/ToneGenerator.kt
package com.spproject.audiowatermark

import kotlin.math.PI
import kotlin.math.sin

/**
 * Converts encrypted bytes into an audio watermark:
 *   preamble tone → 8-bit length header → ciphertext bits (one tone per bit)
 *
 * Wire format (transmitted left-to-right in time):
 *   ┌────────────────┬────────────────────────┬────────────────────────────┐
 *   │  Preamble      │  Length (8 bits)        │  Payload (N×8 bits)        │
 *   │  300 ms        │  8 × 100 ms = 800 ms   │  cipherLen × 8 × 100 ms   │
 *   │  17 000 Hz     │  18k/19.5k Hz per bit  │  18k/19.5k Hz per bit     │
 *   └────────────────┴────────────────────────┴────────────────────────────┘
 *
 * The receiver reads the 8-bit length header to know how many payload bits to
 * collect — no out-of-band parameter is needed on the receiving side.
 */
object ToneGenerator {

    /**
     * Generates a pure sine tone at [freq] Hz for [durationMs] ms.
     * Amplitude is [amplitude] as a fraction of full scale.
     */
    fun generateTone(
        freq: Double,
        durationMs: Int,
        sampleRate: Int = Config.SAMPLE_RATE,
        amplitude: Double = Config.WATERMARK_AMPLITUDE
    ): DoubleArray {
        val numSamples = (sampleRate * durationMs / 1000.0).toInt()
        return DoubleArray(numSamples) { i ->
            amplitude * sin(2.0 * PI * freq * i / sampleRate)
        }
    }

    // ── Bit/byte conversion helpers (public so Decoder can reuse them) ──

    /** Converts bytes to a list of bits, MSB-first, Big Endian. */
    fun bytesToBits(bytes: ByteArray): List<Int> =
        bytes.flatMap { byte -> (7 downTo 0).map { bit -> (byte.toInt() shr bit) and 1 } }

    /**
     * Converts a list of bits (MSB-first) back to bytes.
     * [bits].size must be a multiple of 8; any trailing incomplete byte is ignored.
     */
    fun bitsToBytes(bits: List<Int>): ByteArray {
        val numBytes = bits.size / 8
        return ByteArray(numBytes) { byteIdx ->
            var b = 0
            for (bitPos in 0 until 8) b = (b shl 1) or bits[byteIdx * 8 + bitPos]
            b.toByte()
        }
    }

    // ── Internal helpers ─────────────────────────────────────────────────

    private fun bitsToToneArray(bits: List<Int>): DoubleArray {
        if (bits.isEmpty()) return DoubleArray(0)
        // Allocate the whole output once to avoid repeated array concatenation
        val symbolLen = (Config.SAMPLE_RATE * Config.SYMBOL_DURATION_MS / 1000.0).toInt()
        val result = DoubleArray(bits.size * symbolLen)
        for ((i, bit) in bits.withIndex()) {
            val freq = if (bit == 0) Config.FREQ_BIT_0 else Config.FREQ_BIT_1
            val tone = generateTone(freq, Config.SYMBOL_DURATION_MS)
            tone.copyInto(result, destinationOffset = i * symbolLen)
        }
        return result
    }

    /**
     * Builds the complete self-describing watermark audio sequence:
     *   preamble + 8-bit length header + payload bits
     *
     * [cipherBytes] must be 1–255 bytes (enforced by the 1-byte length header).
     * Returns a [DoubleArray] of normalized [-1.0, 1.0] samples ready to
     * be mixed into the host audio by [Embedder].
     */
    fun buildWatermarkSequence(cipherBytes: ByteArray): DoubleArray {
        require(cipherBytes.isNotEmpty()) { "Ciphertext must not be empty" }
        require(cipherBytes.size <= 255) {
            "Ciphertext is ${cipherBytes.size} bytes; max is 255 (1-byte length header limit). " +
            "Shorten your message."
        }

        val preamble = generateTone(Config.PREAMBLE_FREQ, Config.PREAMBLE_DURATION_MS)

        // 8-bit length header: the number of ciphertext bytes encoded as 8 data tones
        val lengthBits = bytesToBits(byteArrayOf(cipherBytes.size.toByte()))
        val lengthTones = bitsToToneArray(lengthBits)

        // Payload: each ciphertext byte → 8 tones (MSB-first)
        val payloadBits = bytesToBits(cipherBytes)
        val payloadTones = bitsToToneArray(payloadBits)

        // Concatenate: preamble | header | payload
        val totalLen = preamble.size + lengthTones.size + payloadTones.size
        val out = DoubleArray(totalLen)
        preamble.copyInto(out, 0)
        lengthTones.copyInto(out, preamble.size)
        payloadTones.copyInto(out, preamble.size + lengthTones.size)
        return out
    }
}

File: working codes/lib/dsp/tone_generator.dart
import 'dart:math';

import 'protocol.dart';

const int symbolSampleCount = 4410;
const int preambleSampleCount = 13230;

List<double> _generateTone(double frequency, int sampleCount, double amplitude) {
  return List<double>.generate(sampleCount, (index) {
    final phase = 2 * pi * frequency * index / sampleRate;
    return amplitude * sin(phase);
  });
}

List<double> generatePreamble() {
  return _generateTone(preambleFreq, preambleSampleCount, watermarkAmplitude);
}

List<double> bitsToTones(List<int> bits) {
  final tones = <double>[];
  for (final bit in bits) {
    if (bit != 0 && bit != 1) {
      throw ArgumentError.value(bit, 'bit', 'must be 0 or 1');
    }
    final frequency = bit == 0 ? bit0Freq : bit1Freq;
    tones.addAll(_generateTone(frequency, symbolSampleCount, watermarkAmplitude));
  }
  return tones;
}

List<int> bytesToLengthHeaderBits(int length) {
  if (length < 0 || length > 255) {
    throw ArgumentError.value(length, 'length', 'must be between 0 and 255');
  }

  final bits = <int>[];
  for (int repetition = 0; repetition < 3; repetition++) {
    for (int shift = 7; shift >= 0; shift--) {
      bits.add((length >> shift) & 1);
    }
  }
  return bits;
}

List<int> bytesToBits(List<int> bytes) {
  final bits = <int>[];
  for (final byte in bytes) {
    if (byte < 0 || byte > 255) {
      throw ArgumentError.value(byte, 'byte', 'must be between 0 and 255');
    }
    for (int shift = 7; shift >= 0; shift--) {
      bits.add((byte >> shift) & 1);
    }
  }
  return bits;
}

List<int> bitsToBytes(List<int> bits) {
  if (bits.length % 8 != 0) {
    throw ArgumentError.value(bits.length, 'bits.length', 'must be a multiple of 8');
  }
  final bytes = <int>[];
  for (int offset = 0; offset < bits.length; offset += 8) {
    var byte = 0;
    for (int index = 0; index < 8; index++) {
      final bit = bits[offset + index];
      if (bit != 0 && bit != 1) {
        throw ArgumentError.value(bit, 'bit', 'must be 0 or 1');
      }
      byte = (byte << 1) | bit;
    }
    bytes.add(byte);
  }
  return bytes;
}

List<double> encodeFrame(List<int> payloadBits) {
  final payloadLength = payloadBits.length;
  if (payloadLength < 0 || payloadLength > 255) {
    throw ArgumentError.value(payloadLength, 'payloadBits.length', 'must be between 0 and 255');
  }

  final headerBits = bytesToLengthHeaderBits(payloadLength);
  final output = <double>[];
  output.addAll(generatePreamble());
  output.addAll(bitsToTones(headerBits));
  output.addAll(bitsToTones(payloadBits));
  return output;
}

class ToneGenerator {
  static final int _preambleSamples = _sampleCount(preambleDurationMs);
  static final int _symbolSamples = _sampleCount(symbolDurationMs);

  static List<double> generatePreamble() => _generatePreambleImpl();

  static List<double> generateSymbol(int bit) {
    if (bit != 0 && bit != 1) {
      throw ArgumentError.value(bit, 'bit', 'must be 0 or 1');
    }
    final frequency = bit == 0 ? bit0Freq : bit1Freq;
    return _generateTone(frequency, _symbolSamples, watermarkAmplitude);
  }

  static List<double> generatePureTone(
    double freq,
    double durationMs, {
    double? amplitude,
  }) {
    if (freq <= 0) {
      throw ArgumentError.value(freq, 'freq', 'must be greater than 0');
    }
    if (durationMs <= 0) {
      throw ArgumentError.value(durationMs, 'durationMs', 'must be greater than 0');
    }
    final sampleCount = _sampleCount(durationMs);
    final amp = amplitude ?? watermarkAmplitude;
    return List<double>.generate(
      sampleCount,
      (index) => amp * sin(2 * pi * freq * index / sampleRate),
    );
  }

  static List<double> bitsToSamples(List<int> bits) {
    final samples = _generatePreambleImpl();
    for (final bit in bits) {
      samples.addAll(generateSymbol(bit));
    }
    return samples;
  }

  static List<double> bitsToSamplesWithHeader(List<int> payloadBits) {
    if (payloadBits.length % 8 != 0) {
      throw ArgumentError.value(
        payloadBits.length,
        'payloadBits.length',
        'must be a multiple of 8',
      );
    }
    final byteLength = payloadBits.length ~/ 8;
    if (byteLength > 255) {
      throw ArgumentError.value(byteLength, 'byteLength', 'must fit in 8 bits');
    }
    final lengthBits = bytesToBits([byteLength]);
    return bitsToSamples([
      ...lengthBits,
      ...lengthBits,
      ...lengthBits,
      ...payloadBits,
    ]);
  }

  static List<int> bytesToBits(List<int> bytes) => bytesToBits(bytes);

  static List<int> bitsToBytes(List<int> bits) => bitsToBytes(bits);

  static int _sampleCount(double durationMs) {
    return (sampleRate * durationMs / 1000).round();
  }
}

List<double> _generatePreambleImpl() {
  return _generateTone(preambleFreq, preambleSampleCount, watermarkAmplitude);
}
